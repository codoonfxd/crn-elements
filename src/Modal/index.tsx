/*
 * @Author: JaneEyre(lsy@codoon.com)
 * @Date: 2019-04-17 10:56:00
 * @Last Modified by: JaneEyre(lsy@codoon.com)
 * @Last Modified time: 2019-04-17 10:56:00
 * @Content: Modal
 */
import React from 'react'
import {
  View,
  Animated,
  Text,
  TouchableWithoutFeedback,
  ViewStyle,
  TextStyle,
  Easing,
} from 'react-native'
import Button from '../Button'

import { THEME_COLOR, DEVICE_HEIGHT } from '../lib/constant'
import styles from './style'

export interface IModalProps {
  // whether show the modal
  visible: boolean
  // animate type
  animationType: 'none' | 'fade' | 'slide'
  // callback once the modal closed
  onClose: () => void
  // modal title
  title?: string
  // whether the background is transparent
  transparent?: boolean
  // whether mask closing is supported
  maskClosable?: boolean
  // modal animate duration
  animateDuration?: number
  // the bottom button of the modal
  footButtons?: IModalFootItem[]
  // overall style of the modal
  wrapStyle?: ViewStyle
  // mask style of the modal
  maskStyle?: ViewStyle
  // body style of the modal
  style?: ViewStyle
}

export interface IModalFootItem {
  // button text
  text: string
  // press callback
  onPress?: () => void
  // button style
  style?: ViewStyle
  // button text style
  textStyle?: TextStyle
}

interface IState {
  opacity: Animated.Value
  scale: Animated.Value
  visible: boolean
  position: Animated.Value
}

class Modal extends React.Component<IModalProps> {
  static defaultProps: IModalProps = {
    visible: false,
    transparent: true,
    animationType: 'slide',
    // tslint:disable-next-line:no-empty
    onClose: () => {},
    title: '',
    maskClosable: true,
    animateDuration: 100,
    footButtons: [],
    wrapStyle: styles.wrapper as ViewStyle,
    maskStyle: styles.mask as ViewStyle,
    style: styles.modalBody as ViewStyle,
  }

  public state: IState = {
    visible: false,
    scale: new Animated.Value(0),
    opacity: new Animated.Value(0),
    position: new Animated.Value(DEVICE_HEIGHT),
  }

  componentDidUpdate(prevProps: IModalProps) {
    if (prevProps.visible !== this.props.visible) {
      this.toggleShowModal(this.props.visible)
    }
  }

  toggleShowModal = (visible: boolean) => {
    if (visible) {
      this.setState({ visible })
    }
    this.animateModal(visible)
  }

  getPosition = (visible: boolean): number => {
    return visible ? 0 : DEVICE_HEIGHT
  }

  getOpacity = (visible: boolean): number => {
    return visible ? 1 : 0
  }

  getScale = (visible: boolean): number => {
    return visible ? 1 : 0
  }

  animateModal = (visible: boolean): void => {
    const { animationType } = this.props
    let animateAction

    if (animationType === 'fade') {
      animateAction = Animated.parallel([
        Animated.timing(this.state.opacity, {
          toValue: this.getOpacity(visible),
          duration: this.props.animateDuration,
          useNativeDriver: true,
        }),
        Animated.timing(this.state.scale, {
          toValue: this.getScale(visible),
          duration: this.props.animateDuration,
          useNativeDriver: true,
        }),
      ])
    } else if (animationType === 'slide') {
      animateAction = Animated.parallel([
        Animated.timing(this.state.opacity, {
          toValue: this.getOpacity(visible),
          duration: this.props.animateDuration,
          useNativeDriver: true,
        }),
        Animated.timing(this.state.position, {
          useNativeDriver: true,
          easing: Easing.elastic(0.8),
          toValue: this.getPosition(visible),
          duration: this.props.animateDuration,
        }),
      ])
    } else {
      animateAction = Animated.timing(this.state.opacity, {
        useNativeDriver: true,
        toValue: this.getOpacity(visible),
        duration: this.props.animateDuration,
      })
    }
    animateAction.start(() => {
      if (!visible) {
        this.setState({ visible })
      }
    })
  }

  /**
   * the callback after clicking the mask
   */
  onMaskClose = () => {
    const { maskClosable, onClose } = this.props
    if (maskClosable) {
      onClose()
    }
  }

  /**
   * the callback after the bottom button of the modal is pressed
   * @param: foot {object}
   */
  onFootButtonPress = async (foot: IModalFootItem) => {
    if (foot.onPress) {
      await foot.onPress()
    }
    this.props.onClose()
  }

  getFootButtonStyle = (
    foot: IModalFootItem,
    index: number,
    total: number
  ): ViewStyle => {
    const isFirstButton = index === 0
    const isLastButton = index === total - 1
    return {
      ...(styles.footButton as object),
      borderRightWidth: isLastButton ? 0 : 1,
      borderBottomLeftRadius: isFirstButton ? 5 : 0,
      borderBottomRightRadius: isLastButton ? 5 : 0,
      ...foot.style,
    } as ViewStyle
  }

  getFootChild = () => {
    const { footButtons } = this.props
    if (!(footButtons && footButtons.length)) {
      return null
    }
    return (
      <View style={styles.footButtonContaier}>
        {footButtons.map((foot, index) => {
          return (
            <Button
              title={foot.text}
              key={index}
              style={this.getFootButtonStyle(foot, index, footButtons.length)}
              onPress={this.onFootButtonPress.bind(this, foot)}
              underlayColor={'#e1fff3'}
              textStyle={{ color: THEME_COLOR, ...foot.textStyle }}
            />
          )
        })}
      </View>
    )
  }

  render() {
    const { visible } = this.state
    const {
      style,
      title,
      maskStyle,
      wrapStyle,
      transparent,
      animationType,
    } = this.props

    if (!visible) {
      return null
    }

    const animateMap = {
      none: {},
      fade: {
        transform: [{ scale: this.state.scale }],
        opacity: this.state.opacity,
      },
      slide: {
        transform: [{ translateY: this.state.position }],
      },
    }
    return (
      <View style={[styles.wrapper, styles.absolute, wrapStyle]}>
        <TouchableWithoutFeedback onPress={this.onMaskClose}>
          <Animated.View
            style={[
              styles.mask,
              styles.absolute,
              maskStyle,
              transparent ? null : styles.maskWhite,
              {
                opacity: this.state.opacity,
              },
            ]}
          />
        </TouchableWithoutFeedback>
        <Animated.View
          style={[styles.modalBody, style, animateMap[animationType]]}
        >
          {title ? (
            <View style={styles.modalTitleSection}>
              <Text style={styles.modalTitle}>{title}</Text>
            </View>
          ) : null}
          <View style={styles.modalContent}>{this.props.children}</View>
          {this.getFootChild()}
        </Animated.View>
      </View>
    )
  }
}

export default Modal
