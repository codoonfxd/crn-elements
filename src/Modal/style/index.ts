import { StyleSheet } from 'react-native'
import { MODAL_ZINDEX, THEME_COLOR, WHITE_COLOR } from '../../lib/constant'

const styles = StyleSheet.create({
  wrapper: {
    flex: 1,
    zIndex: MODAL_ZINDEX,
    justifyContent: 'center',
    alignItems: 'center',
  },
  absolute: {
    position: 'absolute',
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
  },
  mask: {
    flex: 1,
    backgroundColor: 'rgba(0,0,0,0.2)',
  },
  maskWhite: {
    backgroundColor: WHITE_COLOR,
  },
  modalContent: {
    justifyContent: 'flex-start',
  },
  footButtonContaier: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingVertical: 16,
    paddingHorizontal: 16,
  },
  footCancelButton: {
    flex: 1,
    backgroundColor: WHITE_COLOR,
    borderColor: '#f2f2f2',
    borderWidth: 1,
    borderRadius: 30,
    borderStyle: 'solid',
    marginRight: 16,
  },
  footConfirmButton: {
    flex: 1,
    borderRadius: 35,
    backgroundColor: THEME_COLOR,
  },
  footCancelText: {
    color: '#7d7d7d',
  },
  footConfirmText: {
    color: WHITE_COLOR,
  },
  modalBody: {
    width: 286,
    borderRadius: 5,
    backgroundColor: '#fff',
    flexDirection: 'column',
    justifyContent: 'space-between',
  },
  modalTitleSection: {
    justifyContent: 'center',
    alignItems: 'center',
    paddingVertical: 20,
  },
  modalTitle: {
    color: '#222',
  },
})

export default styles
