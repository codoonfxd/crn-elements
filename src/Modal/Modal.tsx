/*
 * @Author: JaneEyre(lsy@codoon.com)
 * @Date: 2019-04-22 16:25:37
 * @Last Modified by: JaneEyre(lsy@codoon.com)
 * @Last Modified time: 2019-04-22 20:50:20
 * @Content: Modal
 */
import React from 'react'
import {
  View,
  Animated,
  Text,
  TouchableWithoutFeedback,
  ViewStyle,
  TextStyle,
  Easing,
  EasingFunction,
} from 'react-native'
import Button from '../Button'
import ModalStatic from './ModalStatic'

import { DEVICE_HEIGHT } from '../lib/constant'
import styles from './style'
import { Omit } from 'lodash'

export interface IModalProps {
  // whether show the modal
  visible: boolean
  // animate type
  animationType: 'none' | 'fade' | 'slide'
  // callback once the modal closed
  onClose: () => void
  // modal title
  title?: string
  // whether the background is transparent
  transparent?: boolean
  // whether mask closing is supported
  maskClosable?: boolean
  // modal animate duration
  animateDuration?: number
  // animate easing function
  animateEasing?: EasingFunction
  // the bottom button of the modal
  footButtons?: IModalFootItem[]
  // overall style of the modal
  wrapStyle?: ViewStyle
  // mask style of the modal
  maskStyle?: ViewStyle
  // body style of the modal
  style?: ViewStyle
  // foot button style
  footButtonStyle?: ViewStyle
}

export interface IModalFootItem {
  // button text
  text: string
  // press callback
  onPress?: () => void
  // button style
  style?: ViewStyle
  // button text style
  textStyle?: TextStyle
  // button underlayColor
  underlayColor?: string
}

interface IState {
  opacity: Animated.Value
  scale: Animated.Value
  visible: boolean
  position: Animated.Value
}

class Modal extends React.Component<IModalProps, IState> {
  static defaultProps: Omit<IModalProps, 'visible' | 'onClose'> = {
    title: '',
    style: {},
    wrapStyle: {},
    footButtonStyle: {},
    maskStyle: {},
    footButtons: [],
    transparent: true,
    maskClosable: true,
    animateDuration: 200,
    animateEasing: Easing.elastic(0.8),
    animationType: 'fade',
  }

  static alert = ModalStatic.alert
  static prompt = ModalStatic.prompt

  constructor(props: IModalProps) {
    super(props)
    const { visible } = props
    this.state = {
      visible,
      scale: new Animated.Value(this.getScale(visible)),
      opacity: new Animated.Value(this.getOpacity(visible)),
      position: new Animated.Value(this.getPosition(visible)),
    }
  }

  componentDidUpdate(prevProps: IModalProps) {
    if (prevProps.visible !== this.props.visible) {
      this.toggleShowModal(this.props.visible)
    }
  }

  /**
   * toggle the modal display or hide
   * @param { boolean } visible
   */
  toggleShowModal = (visible: boolean) => {
    if (visible) {
      this.setState({ visible })
    }
    this.animateModal(visible)
  }

  /**
   * the callback after clicking the mask
   */
  onMaskClose = () => {
    const { maskClosable, onClose } = this.props
    if (maskClosable) {
      onClose()
    }
  }

  /**
   * the callback after the bottom button of the modal is pressed
   * @param { object } foot
   */
  onFootButtonPress = async (foot: IModalFootItem) => {
    if (foot.onPress) {
      await foot.onPress()
    }
    this.props.onClose()
  }

  /**
   * get animate position value
   * @param { boolean } visible
   * @returns { number }
   */
  getPosition = (visible: boolean): number => {
    return visible ? 0 : DEVICE_HEIGHT
  }

  /**
   * get animate opacity value
   * @param { boolean } visible
   * @returns { number }
   */
  getOpacity = (visible: boolean): number => {
    return visible ? 1 : 0
  }

  /**
   * get animate scale value
   * @param { boolean } visible
   * @returns { number }
   */
  getScale = (visible: boolean): number => {
    return visible ? 1 : 1.06
  }

  /**
   * The method of making the modal move in a preset animation mode
   * @param { boolean } visible
   */
  animateModal = (visible: boolean): void => {
    const { animationType, animateEasing } = this.props
    let animateAction

    if (animationType === 'fade') {
      animateAction = Animated.parallel([
        Animated.timing(this.state.opacity, {
          toValue: this.getOpacity(visible),
          duration: this.props.animateDuration,
          useNativeDriver: true,
        }),
        Animated.timing(this.state.scale, {
          toValue: this.getScale(visible),
          duration: this.props.animateDuration,
          useNativeDriver: true,
        }),
      ])
    } else if (animationType === 'slide') {
      animateAction = Animated.parallel([
        Animated.timing(this.state.opacity, {
          toValue: this.getOpacity(visible),
          duration: this.props.animateDuration,
          useNativeDriver: true,
        }),
        Animated.timing(this.state.position, {
          useNativeDriver: true,
          easing: animateEasing,
          toValue: this.getPosition(visible),
          duration: this.props.animateDuration,
        }),
      ])
    } else {
      animateAction = Animated.timing(this.state.opacity, {
        useNativeDriver: true,
        toValue: this.getOpacity(visible),
        duration: this.props.animateDuration,
      })
    }
    animateAction.start(() => {
      if (!visible) {
        this.setState({ visible })
      }
    })
  }

  /**
   * method to render buttons at the bottom of the modal
   * @param { object } foot
   * @param { number } index
   * @param { number } total
   * @returns { JSX.Element }
   */
  renderFootButtonItem = (
    foot: IModalFootItem,
    index: number,
    total: number
  ) => {
    const isLastButton = index === total - 1
    const buttonStyle: object = isLastButton
      ? styles.footConfirmButton
      : styles.footCancelButton
    const textStyle: object = isLastButton
      ? styles.footConfirmText
      : styles.footCancelText

    return (
      <Button
        key={index}
        title={foot.text}
        underlayColor={foot.underlayColor}
        style={{ ...buttonStyle, ...foot.style } as ViewStyle}
        onPress={this.onFootButtonPress.bind(this, foot)}
        textStyle={{ ...textStyle, ...foot.textStyle } as ViewStyle}
      />
    )
  }

  render() {
    const { visible } = this.state
    const {
      style,
      title,
      maskStyle,
      wrapStyle,
      transparent,
      animationType,
      footButtons,
      footButtonStyle,
    } = this.props

    if (!visible) {
      return null
    }
    const opacity = animationType === 'none' ? 1 : this.state.opacity
    const total = footButtons ? footButtons.length : 0
    const animateMap = {
      none: {},
      fade: {
        transform: [{ scale: this.state.scale }],
        opacity: this.state.opacity,
      },
      slide: {
        transform: [{ translateY: this.state.position }],
      },
    }
    return (
      <View style={[styles.wrapper, styles.absolute, wrapStyle]}>
        <TouchableWithoutFeedback onPress={this.onMaskClose}>
          <Animated.View
            style={[
              styles.mask,
              styles.absolute,
              maskStyle,
              transparent ? null : styles.maskWhite,
              { opacity },
            ]}
          />
        </TouchableWithoutFeedback>
        <Animated.View
          style={[styles.modalBody, style, animateMap[animationType]]}
        >
          {title ? (
            <View style={styles.modalTitleSection}>
              <Text style={styles.modalTitle}>{title}</Text>
            </View>
          ) : null}
          <View style={styles.modalContent}>{this.props.children}</View>
          <View style={[styles.footButtonContaier, footButtonStyle]}>
            {footButtons && footButtons.length
              ? footButtons.map((foot, index) => {
                  return this.renderFootButtonItem(foot, index, total)
                })
              : null}
          </View>
        </Animated.View>
      </View>
    )
  }
}

export default Modal
