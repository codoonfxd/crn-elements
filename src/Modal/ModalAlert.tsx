/*
 * @Author: JaneEyre(lsy@codoon.com)
 * @Date: 2019-04-22 16:05:29
 * @Last Modified by: JaneEyre(lsy@codoon.com)
 * @Last Modified time: 2019-04-22 21:07:21
 * @Content: Modal Alert Component
 */
import * as React from 'react'
import Modal, { IModalFootItem } from './Modal'
import { View, Text, StyleSheet, ViewStyle } from 'react-native'
import { Omit } from 'lodash'

export interface IModalAlertConf {
  title: string
  onClose?: () => any
  content?: string
  footButtons?: IModalFootItem[]
  animateDuration?: number
  contentStyle?: ViewStyle
}

interface IState {
  visible: boolean
}

const styles = StyleSheet.create({
  contentWrapper: {
    marginBottom: 16,
    alignItems: 'center',
    justifyContent: 'center',
  },
  contentStyle: {
    fontSize: 14,
    color: '#9a9a9a',
  },
})

class ModalAlert extends React.Component<IModalAlertConf> {
  static defaultProps: Omit<
    IModalAlertConf,
    'contentStyle' | 'animateDuration' | 'content'
  > = {
    title: '',
    onClose: () => {
      return null
    },
    footButtons: [],
  }

  public state: IState = {
    visible: false,
  }

  componentDidMount() {
    this.setState({ visible: true })
  }

  onClose = () => {
    this.setState({ visible: false })
    const { onClose } = this.props
    if (onClose) {
      onClose()
    }
  }

  render() {
    const {
      title,
      content,
      footButtons,
      animateDuration,
      contentStyle,
    } = this.props
    const { visible } = this.state
    return (
      <Modal
        title={title}
        footButtons={footButtons}
        visible={visible}
        onClose={this.onClose}
        animationType="fade"
        animateDuration={animateDuration}
      >
        {content ? (
          <View style={styles.contentWrapper}>
            <Text style={[styles.contentStyle, contentStyle]}>{content}</Text>
          </View>
        ) : null}
      </Modal>
    )
  }
}

export default ModalAlert
