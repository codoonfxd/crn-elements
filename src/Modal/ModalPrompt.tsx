/*
 * @Author: JaneEyre(lsy@codoon.com)
 * @Date: 2019-04-22 16:24:56
 * @Last Modified by: JaneEyre(lsy@codoon.com)
 * @Last Modified time: 2019-04-22 20:58:27
 * @Content: Modal Prompt Component
 */
import * as React from 'react'
import Modal, { IModalFootItem } from './Modal'
import { ViewStyle, View, Text, StyleSheet, TextInput } from 'react-native'
import { Omit } from 'lodash'

export interface IModalPromptConf {
  title: string
  value: string
  onValueChange: (value: string) => void
  onClose?: () => void
  content?: string
  footButtons?: IModalFootItem[]
  animateDuration?: number
  contentStyle?: ViewStyle
  placeholder?: string
  inputStyle?: ViewStyle
  autoFocus?: boolean
}

interface IState {
  value: string
  visible: boolean
}

const styles = StyleSheet.create({
  contentWrapper: {
    marginBottom: 16,
    alignItems: 'center',
    justifyContent: 'center',
  },
  contentStyle: {
    fontSize: 14,
    color: '#9a9a9a',
  },
  inputStyle: {
    borderWidth: 1,
    borderColor: '#f2f2f2',
    borderStyle: 'solid',
    borderRadius: 2,
    height: 40,
    marginLeft: 16,
    marginRight: 16,
    paddingLeft: 8,
    paddingRight: 8,
  },
})

class ModalPrompt extends React.Component<IModalPromptConf, IState> {
  static defaultProps: Omit<
    IModalPromptConf,
    | 'animateDuration'
    | 'inputStyle'
    | 'contentStyle'
    | 'value'
    | 'onValueChange'
  > = {
    title: '',
    content: '',
    footButtons: [],
    placeholder: '',
    autoFocus: true,
  }

  constructor(props: IModalPromptConf) {
    super(props)
    this.state = {
      visible: false,
      value: props.value,
    }
  }

  componentDidMount() {
    this.setState({ visible: true })
  }

  onValueChange = (value: string) => {
    this.setState({ value })
    this.props.onValueChange(value)
  }

  onClose = () => {
    this.setState({ visible: false })
    const { onClose } = this.props
    if (onClose) {
      onClose()
    }
  }

  render() {
    const { visible, value } = this.state
    const {
      title,
      footButtons,
      animateDuration,
      contentStyle,
      content,
      placeholder,
      inputStyle,
      autoFocus,
    } = this.props
    return (
      <Modal
        visible={visible}
        title={title}
        footButtons={footButtons}
        animationType="fade"
        onClose={this.onClose}
        animateDuration={animateDuration}
      >
        {content ? (
          <View style={styles.contentWrapper}>
            <Text style={[styles.contentStyle, contentStyle]}>{content}</Text>
          </View>
        ) : null}
        <TextInput
          placeholder={placeholder}
          style={[styles.inputStyle, inputStyle]}
          autoFocus={autoFocus}
          value={value}
          onChangeText={this.onValueChange}
        />
      </Modal>
    )
  }
}

export default ModalPrompt
