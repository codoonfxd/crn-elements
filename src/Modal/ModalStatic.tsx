import React from 'react'
import Portal from '../Portal'
import ModalAlert, { IModalAlertConf } from './ModalAlert'

class ModalStatic {
  public key: any
  alert(config: IModalAlertConf) {
    this.key = Portal.add(<ModalAlert {...config} />)
  }
  prompt() {
    console.log(11)
  }
}

export default new ModalStatic()
