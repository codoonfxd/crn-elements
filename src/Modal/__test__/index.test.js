import React from "react";
import { shallow } from 'enzyme'
import renderer from "react-test-renderer";
import Modal from "../Modal";

describe("<Modal />", () => {
  it("snapshot", () => {
    const tree = renderer.create(<Modal />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("render title correctly when the title prop is passed", () => {
    const title = 'test title'
    const wrapper = shallow(<Modal title={title} visible={true} />)
    // expect(wrapper.find(''))
  })  
});
