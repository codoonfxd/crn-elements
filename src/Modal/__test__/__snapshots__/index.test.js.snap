// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> snapshot 1`] = `null`;
